Manage properly the sub-extensions.

MightyCreak opened the pull request:
Factorize code in http/index.php:

* All the functions are defined at the beginning of the file.
* Add functions for extensions so that they can be called by either
  OglVersion or OglExtension.
* Add functions that write HTML code (they all begin with "write").
* The writing of the matrix HTML code is now in writeMatrix().

Add sub-extension paradigm:

* In Parse.php:
    - Create a generateExtension() method that is called by the
      parser or by the extensions.
* In OglExtension.php:
    - Add the sub-extensions array.
    - Add the get and find methods.
    - Add some nice phpDoc comments.
* In OglParser.php:
    - Keep a variable on the last extension parsed.
    - If extension, call OglVersion::addExtension(); if
      sub-extension, call OglExtension::addSubExtension().
* In OglVersion.php:
    - addExtension() now returns the extension "added".
    - Add some nice phpDoc comments.

It's your time to review my code @Xenopathic ;)

Xenopathic commented on lib/Parser/OglVersion.php:
```diff
@@ -63,21 +63,44 @@ public function getGlslVersion() {
+     * @param array $supportedDrivers List of drivers supported for this extension.
+     * @param Mesamatrix\Git\Commit $commit The commit used by the parser.
+     *
+     * @return Mesamatrix\OglExtension The new or existing extension.
```
Put a backslash `\` in front of those namespaced classes, otherwise static code analysers will break.

MightyCreak replied:
Didn't know if it was necessary, but fixed anyway!

Xenopathic commented on lib/Parser/OglExtension.php:
```diff
@@ -120,10 +121,56 @@ public function incorporate($other, $commit) {
+     * @param string $name Name of the extension.
+     * @param string $status Status of the extension.
+     * @param array $supportedDrivers List of drivers supported for this extension.
+     * @param #Mesamatrix\Git\Commit $commit The commit used by the parser.
```
Typo `@param #Mesamatrix...`

MightyCreak replied:
Should I remove `\Mesamatrix\` in here too and keep only `Git\Commit` since it's the same base namespace?

(the same goes for the other lines where I do this kind of reference)

Xenopathic replied:
Unfortunately that doesn't work, PHP (and by extension, the static code analysers) require either a locally-referenced class (aka in the same directory) or an absolute class with full namespace. There is no support for C++ style sub-relative namespaces, as you suggest :smile:

MightyCreak replied:
Ok so the recent changes should be the last one, you confirm?

Xenopathic commented on lib/Parser/OglExtension.php:
```diff
@@ -120,10 +121,56 @@ public function incorporate($other, $commit) {
+     *
+     * @param string $name The name of the extension to find.
+     *
+     * @return \Mesamatrix\OglExtension The extension or null if not found.
```
Wrong namespace - should be `\Mesamatrix\Parser\OglExtension`. But since the namespaces are the same, you can use a relative reference by using just `OglExtension`

Xenopathic commented on lib/Parser/OglVersion.php:
```diff
@@ -63,21 +63,44 @@ public function getGlslVersion() {
+     * @param array $supportedDrivers List of drivers supported for this extension.
+     * @param \Mesamatrix\Git\Commit $commit The commit used by the parser.
+     *
+     * @return \Mesamatrix\OglExtension The new or existing extension.
```
Same here

Xenopathic commented on lib/Parser/OglVersion.php:
```diff
@@ -63,21 +63,44 @@ public function getGlslVersion() {
+     *
+     * @param string $name The name of the extension to find.
+     *
+     * @return \Mesamatrix\OglExtension The extension or null if not found.
```
And here

Xenopathic commented on lib/Parser/OglVersion.php:
```diff
@@ -63,21 +63,44 @@ public function getGlslVersion() {
+    /**
+     * Get the list of all extensions.
+     *
+     * @return array All the extensions.
```
If you use `OglExtension[]` instead of `array` static code checkers will sing praises for you :smile:

MightyCreak commented:
Fixed all the thing you told me to ;)

MightyCreak commented:
Oh I forgot to count the sub-extensions in the total, I'll fix that (but not right now).

MightyCreak commented:
I couldn't resist, I fixed the Leaderboard too ;)
I added it in in the commit message:

* In Leaderboard.php:
    - Need to count by extensions and by sub-extensions.

Xenopathic commented:
Tested, happy with the code, works :+1:
